javaslang/src/main/java/javaslang/concurrent/Future.java

```diff
@@ -802,16 +818,35 @@ default boolean isFailure() {
      * @return A new Future that returns both Future results.
      * @throws NullPointerException if {@code that} is null
      */
-    @SuppressWarnings("unchecked")
     default <U> Future<Tuple2<T, U>> zip(Future<? extends U> that) {
         Objects.requireNonNull(that, "that is null");
-        final Promise<Tuple2<T, U>> promise = Promise.make(executorService());
+        return zipWith(that, Tuple::of);
+    }
+
+    /**
+     * Returns a this and that Future result combined using a given combinator function.
+     * <p>
+     * If this Future failed the result contains this failure. Otherwise the result contains that failure or
+     * a combination of both successful Future results.
+     *
+     * @param that Another Future
+     * @param combinator The combinator function
+     * @param <U>  Result type of {@code that}
+     * @param <R>  Result type of {@code f}
+     * @return A new Future that returns both Future results.
+     * @throws NullPointerException if {@code that} is null
+     */
+    @SuppressWarnings("unchecked")
+    default <U, R> Future<R> zipWith(Future<? extends U> that, Function2<T, U, R> combinator) {
+        Objects.requireNonNull(that, "that is null");
+        Objects.requireNonNull(combinator, "combinator is null");
+        final Promise<R> promise = Promise.make(executorService());
         onComplete(res1 -> {
             if (res1.isFailure()) {
-                promise.complete((Try.Failure<Tuple2<T, U>>) res1);
+                promise.complete((Try.Failure<R>) res1);
             } else {
                 that.onComplete(res2 -> {
-                    final Try<Tuple2<T, U>> result = res1.flatMap(t -> res2.map(u -> Tuple.of(t, u)));
+                    final Try<R> result = res1.flatMap(t -> res2.map(u -> combinator.apply(t, u)));
```

Comment:
Mmh, please add a unit test that let's the combinator fail, e.g. a `combinator` that throws a `RuntimeException`.

I expect that the `Future` that is returned by `zipWith` does never return.

**Note to myself:** We need also to check the other Future/FutureImpl methods. It must be ensured that in each case we get a defined result. We can do it by adding tests that use lambda arguments that just throw a RuntimeException. `map` delegated to `mapTry`, which should handle that case. The current optimization may change this behavior. #LetItCrash